### PR TITLE
[SPARK-13990] Automatically pick serializer when caching RDDs

### PR DESCRIPTION
**Pull request by JoshRosen**
Building on the `SerializerManager` introduced in SPARK-13926/ #11755, this patch Spark modifies Spark's BlockManager to use RDD's ClassTags in order to select the best serializer to use when caching RDD blocks.

When storing a local block, the BlockManager `put()` methods use implicits to record ClassTags and stores those tags in the blocks' BlockInfo records. When reading a local block, the stored ClassTag is used to pick the appropriate serializer. When a block is stored with replication, the class tag is written into the block transfer metadata and will also be stored in the remote BlockManager.

There are two or three places where we don't properly pass ClassTags, including TorrentBroadcast and BlockRDD. I think this happens to work because the missing ClassTag always happens to be `ClassTag.Any`, but it might be worth looking more carefully at those places to see whether we should be more explicit.

### CODE REVIEW

**Comment by JoshRosen** [comment]
```
[error]  * method uploadBlock(java.lang.String,Int,java.lang.String,org.apache.spark.storage.BlockId,org.apache.spark.network.buffer.ManagedBuffer,org.apache.spark.storage.StorageLevel)scala.concurrent.Future in class org.apache.spark.network.netty.NettyBlockTransferService does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.network.netty.NettyBlockTransferService.uploadBlock")
```

@rxin is it okay to change this?

**Comment by rxin** [comment]
Yes - that should be an internal module.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -703,13 +710,13 @@ private[spark] class BlockManager(
    *
    * @return true if the block was stored or false if an error occurred.
    */
-  def putBytes(
+  def putBytes[T: ClassTag]( // TODO(josh)
```
Comment (nongli) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: is this adddressed?

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Yes. During the original design I wasn't sure whether I wanted this to be an implicit ClassTag, but in this case it's actually desired / fine.

**Comment by nongli** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/network/BlockDataManager.scala [hunk 2]
```diff
@@ -35,7 +37,11 @@ trait BlockDataManager {
    * Returns true if the block was stored and false if the put operation failed or the block
    * already existed.
    */
-  def putBlockData(blockId: BlockId, data: ManagedBuffer, level: StorageLevel): Boolean
+  def putBlockData(
+      blockId: BlockId,
+      data: ManagedBuffer,
+      level: StorageLevel,
+     classTag: ClassTag[_]): Boolean
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/network/BlockDataManager.scala]: indentation

**Comment by JoshRosen** [comment]
Alright, I'm going to merge this now and will address any problems in followups.

**Pull request closed without merging**